content/post/24-Flux-iOS.md

```diff
@@ -0,0 +1,515 @@
+
+State management was one of the largest sources of complexity in our app, so we decided to define a pattern that all new features could use going forward.
+
+A lot of the pain we felt in our existing codebase reminded us strongly of the issues that Facebook brought up when they first presented the Flux pattern:
+
+- Unexpected, Cascading State Updates
+- Hard to Understand Dependencies Between Components
+- Tangled Flow of Information
+- Unclear Source of Truth
+
+It seemed that Flux would be a great fit for solving many of the issues we were experiencing.
+
+## A Brief Intro to Flux
+
+Flux is a very lightweight architectural pattern that Facebook uses for client-side web applications. Even though there is a [reference implementation](https://github.com/facebook/flux), Facebook emphasizes that the ideas of the Flux pattern are a lot more relevant than this particular implementation.
+
+The pattern can be described best alongside a diagram that shows the different flux components:
+
+![](https://raw.githubusercontent.com/Ben-G/Website/flux-post/static/assets/flux-post/Flux_Original.png)
+
+In the Flux architecture a **store** is the single source of truth for a certain part of the app. Whenever the state in the store updates, it will call a handler method on all views that subscribed to the store. The **view** receives state updates only through this one interface that is called by the store.
+
+State updates can only occurr via **actions**.
+
+An **action** describes an intented state change, but it doesn't implement the state change itself. All components that want to change any state send an **action** to the global **dispatcher**.
+The stores register with the dispatcher and let it know which actions they are interested in. Whenever an action is dispatched, all interested stores will receive it.
+
+In response to actions some stores will update their state and notify the views about that new state.
+
+The Flux architecture enforces a unidirectional data flow as shown in the diagram above. It also enforces a strict separation of concerns:
+
+- Views will only receive data from stores. Whenever a store updates, the handler method on the view is invoked.
+- Views can only change state by dispatching actions. Because actions are only descriptions of intents, the business logic is hidden from the view.
+- A store only updates its state when it receives an action.
+
+These constraints make designing, developing and debugging new features a lot easier.
+
+## Flux in PlanGrid for iOS
+
+For the PlanGrid iOS app we have extended the Flux specification a little bit. We enforce that each store needs to have an explicit state type. This state type describes the entire state for a certain feature. This state is observed by our views and is used to update them whenever a state change occurs:
```

Comment:
@zats I describe that further along with an example of the state type for the annotation filter. However, I agree I should elaborate on this. I should probably describe this slightly different.
Instead of sending a change event the way the original Flux architecture does, we provide an observable state property.